File: gachisa-backend/src/main/java/com/gachisa/payment/service/RefundService.java
package com.gachisa.payment.service;

import com.gachisa.global.exception.CustomException;
import com.gachisa.global.exception.ErrorCode;
import com.gachisa.payment.client.PgClient;
import com.gachisa.payment.client.dto.PgCancellationResult;
import com.gachisa.payment.dto.RefundResponse;
import com.gachisa.payment.service.dto.RefundPreparation;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class RefundService {

    private final RefundStateService refundStateService;
    private final RefundCompletionService refundCompletionService;
    private final PgClient pgClient;

    public RefundResponse refund(Long paymentId, String reason) {
        RefundResponse requested = requestRefund(paymentId, reason);
        return processPending(requested.refundId());
    }

    public RefundResponse requestRefund(Long paymentId, String reason) {
        RefundPreparation preparation = refundStateService.prepare(paymentId, reason);
        return refundStateService.getRefund(preparation.refundId());
    }

    public RefundResponse processPending(Long refundId) {
        RefundPreparation preparation = refundStateService.claimPending(refundId);
        if (!preparation.requestRequired()) {
            return refundStateService.getRefund(refundId);
        }

        try {
            PgCancellationResult result = pgClient.cancel(
                    preparation.paymentKey(),
                    preparation.reason(),
                    preparation.pgIdempotencyKey()
            );
            return refundCompletionService.complete(preparation.refundId(), result);
        } catch (CustomException exception) {
            if (exception.getErrorCode() == ErrorCode.PAYMENT_GATEWAY_REJECTED) {
                refundStateService.fail(preparation.refundId(), exception.getErrorCode());
            } else {
                refundStateService.keepPending(preparation.refundId(), exception.getErrorCode());
            }
            throw exception;
        }
    }
}

File: gachisa-backend/src/main/kotlin/com/gachisa/payment/service/RefundService.kt
package com.gachisa.payment.service

import com.gachisa.global.exception.CustomException
import com.gachisa.global.exception.ErrorCode
import com.gachisa.payment.client.PgClient
import com.gachisa.payment.dto.RefundResponse
import com.gachisa.payment.metric.PaymentMetrics
import org.springframework.stereotype.Service

@Service
class RefundService(
    private val refundStateService: RefundStateService,
    private val refundCompletionService: RefundCompletionService,
    private val pgClient: PgClient,
    private val paymentMetrics: PaymentMetrics,
) {
    fun refund(paymentId: Long, reason: String): RefundResponse {
        val requested = requestRefund(paymentId, reason)
        return processPending(requested.refundId)
    }

    fun requestRefund(paymentId: Long, reason: String): RefundResponse {
        val preparation = refundStateService.prepare(paymentId, reason)
        paymentMetrics.recordRefund(if (preparation.requestRequired) "requested" else "idempotent")
        return refundStateService.getRefund(preparation.refundId)
    }

    fun processPending(refundId: Long): RefundResponse {
        val preparation = refundStateService.claimPending(refundId)
        if (!preparation.requestRequired) return refundStateService.getRefund(refundId)

        try {
            val result = paymentMetrics.recordRefundTime { pgClient.cancel(
                preparation.paymentKey!!,
                preparation.reason,
                preparation.pgIdempotencyKey,
            ) }
            val response = refundCompletionService.complete(preparation.refundId, result)
            paymentMetrics.recordRefund("success")
            return response
        } catch (exception: CustomException) {
            if (exception.getErrorCode() == ErrorCode.PAYMENT_GATEWAY_REJECTED) {
                refundStateService.fail(preparation.refundId, exception.getErrorCode())
                paymentMetrics.recordRefund("failed")
            } else {
                refundStateService.keepPending(preparation.refundId, exception.getErrorCode())
                paymentMetrics.recordRefund("retry_scheduled")
            }
            throw exception
        }
    }
}
